### PR TITLE
[BUGFIX beta] References load() enhancements

### PR DESCRIPTION
**Pull request by tchak**
- ensure reference.load() is always returning a simple RSVP.Promise not an ObjectProxy

I don't expect any breaking changes here.

### CODE REVIEW

**Comment by tchak** [comment]
@igorT do you see any problems with exposing `isLoaded()` on the references?

File: addon/-private/system/references/belongs-to.js [hunk 1]
```diff
@@ -63,32 +63,33 @@ BelongsToReference.prototype.push = function(objectOrPromise) {
   });
 };
 
-BelongsToReference.prototype.value = function() {
-  var inverseRecord = this.belongsToRelationship.inverseRecord;
+BelongsToReference.prototype.isLoaded = function() {
+  let inverseRecord = this.belongsToRelationship.inverseRecord;
+
+  return inverseRecord && inverseRecord.isLoaded();
```
Comment (igorT) [inline addon/-private/system/references/belongs-to.js]: This seems to miss the case where it is null. If we know we are null, are we loaded?

File: addon/-private/system/references/belongs-to.js [hunk 2]
```diff
@@ -63,32 +63,33 @@ BelongsToReference.prototype.push = function(objectOrPromise) {
...
   }
 
   return null;
 };
 
 BelongsToReference.prototype.load = function() {
-  if (this.remoteType() === "id") {
-    return this.belongsToRelationship.getRecord();
-  }
+  let record = this.value();
```
Comment (igorT) [inline addon/-private/system/references/belongs-to.js]: Does this cache get wiped correctly?

File: addon/-private/system/references/has-many.js [hunk 3]
```diff
@@ -121,24 +121,26 @@ HasManyReference.prototype._isLoaded = function() {
...
   }
 
-  var manyArray = this.hasManyRelationship.manyArray;
   return Ember.RSVP.resolve(manyArray);
 };
 
 HasManyReference.prototype.reload = function() {
-  return this.hasManyRelationship.reload();
+  return this.hasManyRelationship.reload()
+    .then(() => this.value());
```
Comment (igorT) [inline addon/-private/system/references/has-many.js]: This seems to swallow the error case?

Comment (fivetanley) [inline addon/-private/system/references/has-many.js]: should still reject with the error.

File: addon/-private/system/references/record.js [hunk 4]
```diff
@@ -26,18 +26,25 @@ RecordReference.prototype.push = function(objectOrPromise) {
...
+  if (this.isLoaded()) {
+    return this.internalModel.getRecord();
+  }
+
+  return null;
 };
 
 RecordReference.prototype.load = function() {
-  return this.store.findRecord(this.type, this._id);
+  return Ember.RSVP.resolve(this.store.findRecord(this.type, this._id));
```
Comment (igorT) [inline addon/-private/system/references/record.js]: What if the load fails?

File: tests/integration/references/belongs-to-test.js [hunk 5]
```diff
@@ -230,7 +230,7 @@ test("push(promise)", function(assert) {
     push = familyReference.push(deferred.promise);
   });
 
-  assert.ok(push.then, 'BelongsToReference.push returns a promise');
+  assert.ok(push instanceof Ember.RSVP.Promise, 'BelongsToReference.push() returns a promise');
```
Comment (igorT) [inline tests/integration/references/belongs-to-test.js]: I think the official promise spec is to check for a .then.

**Comment by igorT** [comment]
@tchak the semantics are a bit tricky with relationships that are not links. If you keep foreign keys on the belongs to side, how can you know if the has many has been loaded. We have an internal `hasData` for a lot of these purposes which @wecc wrote, might be useful to look at it

File: tests/integration/references/belongs-to-test.js (hunk 5 above)
Comment (tchak) [inline tests/integration/references/belongs-to-test.js]: yes but what I am trying to test is that we are not returning `PromiseProxyObject`. If you have a better idea on how to test this...

Comment (fivetanley) [inline tests/integration/references/belongs-to-test.js]: maybe `! instanceof PromiseProxyObject` ? might be the most clear, as well as updating the description.

File: addon/-private/system/references/belongs-to.js (hunk 2 above)
Comment (tchak) [inline addon/-private/system/references/belongs-to.js]: What cache are you talking about?

File: tests/integration/references/belongs-to-test.js (hunk 5 above)
Comment (tchak) [inline tests/integration/references/belongs-to-test.js]: I changed, but I feel much less confident in the tests this way...

File: addon/-private/system/references/record.js (hunk 4 above)
Comment (tchak) [inline addon/-private/system/references/record.js]: `Ember.RSVP.resolve(Ember.RSVP.reject("yolo")).catch(console.log.bind(console)) // => "yolo"`

But I added some tests to be sure :)

File: addon/-private/system/references/belongs-to.js (hunk 2 above)
Comment (igorT) [inline addon/-private/system/references/belongs-to.js]: Actually this seems weird. So first you do `this.value()` which is  `this.belongsToRelationship.inverseRecord.getRecord()`  but then if it's null you do `this.belongsToRelationship.getRecord();` Do we just have two different things named `getRecord`? One on intneralModel and the other on the relationship? Maybe we should rename?

File: addon/-private/system/references/has-many.js (hunk 3 above)
Comment (tchak) [inline addon/-private/system/references/has-many.js]: And it will reject with errors. I am only overriding resolved result. I added some tests to prouve me right :)

File: addon/-private/system/references/belongs-to.js (hunk 2 above)
Comment (tchak) [inline addon/-private/system/references/belongs-to.js]: yes, we do. `inverseRecord` is actually an instance of `internalModel`. `getRecord` is the correct way of getting, well a record, from an `internalModel`. `getRecord` on the `BelongsToRelationship` is method to fetch a record (it deals internally with links and stuff).

**Comment by tchak** [comment]
@igorT yes we already check `hasData` in the `has-many-reference#isLoaded`. I added the same check to `belongs-to-reference#isLoaded`.

File: addon/-private/system/store/finders.js [hunk 6]
```diff
@@ -44,9 +44,6 @@ export function _find(adapter, store, typeClass, id, internalModel, options) {
     });
   }, function(error) {
     internalModel.notFound();
-    if (internalModel.isEmpty()) {
```
Comment (tchak) [inline addon/-private/system/store/finders.js]: @igorT do you remember why we need this?

File: addon/-private/system/store/finders.js [hunk 7]
```diff
@@ -80,7 +77,13 @@ export function _findMany(adapter, store, typeClass, ids, internalModels) {
 
       return internalModels;
     });
-  }, null, "DS: Extract payload of " + typeClass);
+  }, function(error) {
+    for (let i = 0; i < internalModels.length; i++) {
```
Comment (tchak) [inline addon/-private/system/store/finders.js]: This basically mimics what we already doing in the `_find` method.

**Comment by tchak** [comment]
@igorT @fivetanley I added a bunch of tests for failure/retry cases. Bunch of stuff were broken (before my refactoring). I have fixes for all of the broken stuff but it involves some changes to what looks to be non tested edge cases.

**Comment by tchak** [comment]
I think this PR also fix #4146

File: addon/-private/system/references/belongs-to.js [hunk 8]
```diff
@@ -5,6 +5,9 @@ import Reference from './reference';
 import isEnabled from '../../features';
 import { assertPolymorphicType, deprecate } from "ember-data/-private/debug";
 
+const {
+  RSVP: { resolve }
```
Comment (igorT) [inline addon/-private/system/references/belongs-to.js]: I think we don't do this

Comment (igorT) [inline addon/-private/system/references/belongs-to.js]: The destructuring i mean

Comment (hjdivad) [inline addon/-private/system/references/belongs-to.js]: What's the reason not to?  It's the array destructuring that's dangerous.

File: tests/integration/references/belongs-to-test.js [hunk 9]
```diff
@@ -405,13 +406,62 @@ test("value() returns the referenced record when loaded", function(assert) {
   assert.equal(familyReference.value(), family);
 });
 
+test("value() is null when reference is loading", function(assert) {
+  var done = assert.async();
+  var deferred = Ember.RSVP.defer();
```
Comment (hjdivad) [inline tests/integration/references/belongs-to-test.js]: `var`?

File: tests/integration/references/belongs-to-test.js [hunk 10]
```diff
@@ -405,13 +406,62 @@ test("value() returns the referenced record when loaded", function(assert) {
...
+
+  assert.equal(familyReference.value(), null);
+
+  deferred.resolve({
+    id: 1, name: "Coreleone"
+  });
+
+  load.then(family => {
+    assert.equal(familyReference.value(), family);
+    done();
```
Comment (hjdivad) [inline tests/integration/references/belongs-to-test.js]: we can just `return` the promise rather than using `assert.async`

**Comment by runspired** [comment]
@tchak I really like this, can we reboot it?

**Comment by runspired** [comment]
I think we should reboot this, but I'm closing this attempt in favor of a new attempt that takes into consideration `RecordData`.
